File: grimdb/sdk/dart/lib/src/file_vault/file_vault_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import '../grimlocker_exception.dart';
import '../models/vault_entry.dart';

class FileVaultClient {
  final String _endpoint;
  final Map<String, String> _headers;
  final http.Client _http;

  FileVaultClient(this._endpoint, this._headers, this._http);

  // ── Circuit breaker state ─────────────────────────────────────────────────────

  int _consecutiveFailures = 0;
  DateTime? _circuitOpenUntil;
  bool _isProbe = false;

  void _onFailure() {
    if (_isProbe) {
      _circuitOpenUntil = DateTime.now().add(const Duration(seconds: 30));
      _isProbe = false;
      return;
    }
    _consecutiveFailures++;
    if (_consecutiveFailures >= 5) {
      _circuitOpenUntil = DateTime.now().add(const Duration(seconds: 30));
      _consecutiveFailures = 0;
    }
  }

  void _onSuccess() {
    _consecutiveFailures = 0;
    _circuitOpenUntil = null;
    _isProbe = false;
  }

  Future<FolderListing> listFolder({String folderId = ''}) async {
    final result = await _call('file.list_folder', {'folder_id': folderId});
    return FolderListing.fromJson(result);
  }

  Future<FolderItem> createFolder(String name, {String parentId = ''}) async {
    final result = await _call(
        'file.create_folder', {'name': name, 'parent_id': parentId});
    return FolderItem.fromJson(result);
  }

  Future<void> renameFolder(String id, String name) async {
    await _call('file.rename_folder', {'id': id, 'name': name});
  }

  Future<void> deleteFolder(String id) async {
    await _call('file.delete_folder', {'id': id});
  }

  Future<void> moveFile(String manifestBlockId, String folderId) async {
    await _call('file.move', {
      'manifest_block_id': manifestBlockId,
      'folder_id': folderId,
    });
  }

  Future<FileEntry> uploadFile(
    Uint8List data,
    String fileName, {
    String mimeType = 'application/octet-stream',
    String folderId = '',
    void Function(UploadProgress)? onProgress,
  }) async {
    onProgress?.call(UploadProgress(bytesSent: 0, totalBytes: data.length));
    final dataB64 = base64Encode(data);
    final result = await _call('file.ingest', {
      'file_name': fileName,
      'mime_type': mimeType,
      'folder_id': folderId,
      'data_b64': dataB64,
    });
    onProgress?.call(UploadProgress(bytesSent: data.length, totalBytes: data.length));
    return FileEntry.fromJson(result);
  }

  Future<Uint8List> downloadFile(String manifestBlockId) async {
    final result = await _call('file.download', {
      'manifest_block_id': manifestBlockId,
    });
    final dataB64 = result['data_b64'] as String?;
    if (dataB64 == null) {
      throw GrimlockerException(-10, 'Download returned no data');
    }
    return base64Decode(dataB64);
  }

  Future<dynamic> _call(
      String action, Map<String, dynamic> payload) async {
    if (_circuitOpenUntil != null) {
      if (DateTime.now().isBefore(_circuitOpenUntil!)) {
        throw CircuitBreakerOpenException();
      }
      _isProbe = true;
    }

    int attempt = 0;
    int delayMs = 100;
    Exception? lastException;

    while (true) {
      try {
        final body = jsonEncode({'action': action, 'payload': payload});
        final response = await _http.post(
          Uri.parse(_endpoint),
          headers: _headers,
          body: body,
        );

        final responseBody = response.body;
        dynamic json;

        try {
          json = jsonDecode(responseBody);
        } catch (_) {
          json = {'error': responseBody};
        }

        if (response.statusCode >= 400 && response.statusCode < 500) {
          final map = json is Map<String, dynamic> ? json : <String, dynamic>{};
          final code = map['error_code'] as int? ?? 0;
          final msg = map['error'] as String? ?? responseBody;
          throw GrimlockerException(
              code, '${GrimlockerException.nameOf(code)} ($action): $msg');
        }

        if (response.statusCode < 200 || response.statusCode >= 300) {
          final map = json is Map<String, dynamic> ? json : <String, dynamic>{};
          final code = map['error_code'] as int? ?? 0;
          final msg = map['error'] as String? ?? responseBody;
          lastException = GrimlockerException(
              code, '${GrimlockerException.nameOf(code)} ($action): $msg');
        } else {
          _onSuccess();
          if (json is Map<String, dynamic>) return json;
          if (json is List) return json;
          return <String, dynamic>{};
        }
      } on GrimlockerException {
        _onFailure();
        rethrow;
      } on http.ClientException catch (e) {
        lastException = e;
      } on TimeoutException catch (e) {
        lastException = e;
      } catch (e) {
        _onFailure();
        rethrow;
      }

      if (attempt >= 3) break;
      await Future.delayed(Duration(milliseconds: min(delayMs, 2000)));
      delayMs *= 2;
      attempt++;
    }

    _onFailure();
    throw lastException ?? Exception('Request failed after retries');
  }
}

File: grimdb/sdk/kotlin/src/main/kotlin/com/grimlocker/sdk/FileVaultClient.kt
package com.grimlocker.sdk

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.grimlocker.sdk.model.*
import java.io.OutputStream
import java.lang.reflect.Type
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64

class FileVaultClient(
    private val baseUrl: String,
    private val token: String,
    private val timeoutMs: Int = 30_000
) {
    private val gson = Gson()
    private val apiUrl: String = "${baseUrl.trimEnd('/')}/api/v1"

    // ── Circuit breaker state ─────────────────────────────────────────────────

    private var consecutiveFailures = 0
    private var circuitOpenUntil: Long = 0
    private var isProbe = false

    private fun onSuccess() {
        consecutiveFailures = 0
        circuitOpenUntil = 0
        isProbe = false
    }

    private fun onFailure() {
        if (isProbe) {
            circuitOpenUntil = System.currentTimeMillis() + 30_000
            isProbe = false
            return
        }
        consecutiveFailures++
        if (consecutiveFailures >= 5) {
            circuitOpenUntil = System.currentTimeMillis() + 30_000
            consecutiveFailures = 0
        }
    }

    fun listFolder(folderId: String = ""): FolderListing =
        callSingle("file.list_folder", mapOf("folder_id" to folderId))

    fun createFolder(name: String, parentId: String = ""): FolderItem =
        callSingle("file.create_folder", mapOf("name" to name, "parent_id" to parentId))

    fun renameFolder(id: String, name: String) {
        call("file.rename_folder", mapOf("id" to id, "name" to name))
    }

    fun deleteFolder(id: String) {
        call("file.delete_folder", mapOf("id" to id))
    }

    fun moveFile(manifestBlockId: String, folderId: String) {
        call("file.move", mapOf("manifest_block_id" to manifestBlockId, "folder_id" to folderId))
    }

    fun uploadFile(
        data: ByteArray,
        fileName: String,
        mimeType: String = "application/octet-stream",
        folderId: String = "",
        onProgress: ((UploadProgress) -> Unit)? = null
    ): FileEntry {
        onProgress?.invoke(UploadProgress(bytesSent = 0, totalBytes = data.size.toLong()))
        val dataB64 = Base64.getEncoder().encodeToString(data)
        val entry: FileEntry = callSingle(
            "file.ingest",
            mapOf(
                "file_name" to fileName,
                "mime_type" to mimeType,
                "folder_id" to folderId,
                "data_b64" to dataB64
            )
        )
        onProgress?.invoke(UploadProgress(bytesSent = data.size.toLong(), totalBytes = data.size.toLong()))
        return entry
    }

    fun downloadFile(manifestBlockId: String): ByteArray {
        data class DownloadResponse(val data_b64: String?)
        val resp: DownloadResponse = callSingle(
            "file.download",
            mapOf("manifest_block_id" to manifestBlockId)
        )
        val b64 = resp.data_b64
            ?: throw GrimlockerException("Download returned no data")
        return Base64.getDecoder().decode(b64)
    }

    private fun call(action: String, payload: Map<String, Any?>) {
        doPost(action, payload)
    }

    private inline fun <reified T> callSingle(action: String, payload: Map<String, Any?>): T {
        val body = doPost(action, payload)
        return try {
            gson.fromJson(body, T::class.java)
        } catch (e: GrimlockerException) {
            throw e
        } catch (e: Exception) {
            throw GrimlockerException(-100, "Failed to parse response: ${e.message}")
        }
    }

    private fun doPost(action: String, payload: Map<String, Any?>): String {
        if (circuitOpenUntil > 0) {
            if (System.currentTimeMillis() < circuitOpenUntil) {
                throw CircuitBreakerOpenException()
            }
            isProbe = true
        }

        var attempt = 0
        var delayMs = 100L
        var lastException: Exception? = null

        while (true) {
            try {
                val requestBody = mapOf("action" to action, "payload" to payload)
                val json = gson.toJson(requestBody)

                val url = URL(apiUrl)
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    doOutput = true
                    connectTimeout = timeoutMs
                    readTimeout = timeoutMs
                    setRequestProperty("Content-Type", "application/json")
                    setRequestProperty("Accept", "application/json")
                    setRequestProperty("X-Grimlocker-Token", token)
                    setRequestProperty("User-Agent", "GrimlockerSDK-Kotlin/1.0.0")
                }

                try {
                    conn.outputStream.use { os: OutputStream ->
                        os.write(json.toByteArray(Charsets.UTF_8))
                        os.flush()
                    }

                    val status = conn.responseCode

                    if (status in 200..299) {
                        val body = conn.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
                        onSuccess()
                        return body
                    }

                    val errorBody = try {
                        conn.errorStream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() } ?: ""
                    } catch (e: Exception) {
                        ""
                    }

                    if (status in 400..499) {
                        parseAndThrow(status, errorBody)
                    }

                    lastException = GrimlockerException("HTTP $status: $errorBody")
                } finally {
                    conn.disconnect()
                }
            } catch (e: GrimlockerException) {
                onFailure()
                throw e
            } catch (e: java.net.SocketException) {
                lastException = e
            } catch (e: java.net.UnknownHostException) {
                lastException = e
            } catch (e: java.net.ConnectException) {
                lastException = e
            } catch (e: java.net.SocketTimeoutException) {
                lastException = e
            } catch (e: Exception) {
                onFailure()
                throw e
            }

            if (attempt >= 3) break
            Thread.sleep(delayMs.coerceAtMost(2000))
            delayMs *= 2
            attempt++
        }

        onFailure()
        throw lastException ?: Exception("Request failed after retries")
    }

    private fun parseAndThrow(httpStatus: Int, body: String): Nothing {
        val code: Int
        val message: String

        try {
            val errorMap = gson.fromJson<Map<String, Any?>>(
                body,
                object : TypeToken<Map<String, Any?>>() {}.type
            )
            code = (errorMap["error_code"] as? Double)?.toInt() ?: 0
            message = (errorMap["error"] as? String) ?: body.ifEmpty { "HTTP $httpStatus" }
        } catch (e: Exception) {
            code = 0
            message = body.ifEmpty { "HTTP $httpStatus" }
        }

        throw GrimlockerException(code, message)
    }
}
